app/models/host_facets/infrastructure_facet.rb

```diff
@@ -0,0 +1,30 @@
+module HostFacets
+  class InfrastructureFacet < Base
+    belongs_to :smart_proxy
+
+    def self.populate_fields_from_facts(host, parser, type, source_proxy)
+      # Do not do anything if there are no relevant facts
+      return unless parser.foreman_uuid || parser.smart_proxy_uuid
+
+      facet = host.infrastructure_facet || host.build_infrastructure_facet
+      facet.foreman_uuid = parser.foreman_uuid
+      facet.smart_proxy_uuid = parser.smart_proxy_uuid
+      facet.refresh_associations!
+      facet.save if facet.changed?
```

Comment:
This is the most important method in regard to performance.

Do you really need to refresh associations even when factet was not changed? That looks fishy, it could lead to updating smart proxy id on every fact request (for every single host) putting unwated stress on the db.

Reply:
`InfrastructureFacet#refresh_associations!` checks whether the proxy uuid changed and only then refreshes the ids https://github.com/theforeman/foreman/pull/8254/files#diff-a8f5825bac3a5e32b754ccbd227a581c80dc1d0a9ffb0a07cdd0f7c0af9a4383R24-R28

Should I move the check out of there and only call it if something changed?

Reply:
Then it is fine, I have missed the method actually does nothing. But I would not call it if not needed - someone can change it later on and we can generate unnecessary calls. Up to you.